### ISSUE TITLE
Embeddable Properties Incorrectly Marked as Optional

### ISSUE THREAD

**Issue by devcaeg** [mikro-orm/mikro-orm#4748]
**Describe the bug**
There's an issue with the type definitions in MikroOrm. When you have an entity that includes an "Embeddable", TypeScript marks all properties of that Embeddable as optional. This leads to entities being created without requesting parameters that are actually required.

**Example**
```typescript
@Embeddable()
export class UserEmail {
  @Property({
    hidden: true,
  })
  code!: number;

  @Property()
  codeSentAt?: Date;

  @Enum({
    items: () => UserEmailStatuses,
  })
  status!: UserEmailStatuses;

  @Property({
    unique: true,
  })
  value!: string;
}

@Embeddable()
export class UserPhoneMobile {
  @Property({
    hidden: true,
  })
  code!: number;

  @Property()
  codeSentAt?: Date;

  @ManyToOne()
  country!: Country;

  @Property()
  number!: string;

  @Enum({
    items: () => UserPhoneMobileStatuses,
  })
  status!: UserPhoneMobileStatuses;
}

@Embeddable()
export class UserPhone {
  @Embedded({
    object: true,
  })
  mobile!: UserPhoneMobile;
}

@Entity({
  collection: 'users',
})
export class User extends Base {
  @Embedded({
    object: true,
  })
  email?: UserEmail;

  @Embedded({
    object: true,
  })
  phone!: UserPhone;

  @Property({
    hidden: true,
  })
  password?: string;

  @Property({
    unique: true,
  })
  username?: string;
}
```
If you try to create an entity named "User", you'll notice that all the properties of the "Embeddable UserEmail" appear as optional. Yet, it has required properties such as "code", "status", and "value". The same situation arises with the "Embeddable UserPhone": the "mobile" property is required, but when you create the "User" entity, "mobile" is displayed as optional, as are all properties contained within "mobile".

**Expected behavior**
Optional properties should remain optional, and required ones should stay required.

**Additional context**
I'm using the metadata provider "TsMorphMetadataProvider".

**Versions**
| Dependency | Version |
| - | - |
| node | 20.5.1 |
| typescript | 5.2.2 |
| mikro-orm | 5.8.3 |
| mikro-orm/mongodb | 5.8.3 |
| nestjs | 10.2.6 |

**Comment by B4nan** [comment]
Fixing this is actually pretty breaking (on type level), so will leave that for v6.

**Comment by devcaeg** [comment]
> Fixing this is actually pretty breaking (on type level), so will leave that for v6. The problem is there for all relations, not just embedded properties.

Sad, but understandable, is there an approximate release date for version 6?

**Comment by B4nan** [comment]
End of the year, but hopefully sooner.

**Comment by B4nan** [comment]
But you can use v6 already, every commit is published to npm, upgrading guide is there too. It's just not stable, some minor breaking changes might be still on the way (although I don't have any planned right now).

**Comment by devcaeg** [comment]
> But you can use v6 already, every commit is published to npm, upgrading guide is there too. It's just not stable, some minor breaking changes might be still on the way (although I don't have any planned right now).

Excellent, thank you very much for the information.

**Comment by B4nan** [comment]
There is one common issue with this, and I think I found a way to make it less dumb, although it's just a quirk. If we start enforcing this on relations, the `em.create` automatically propagates things to inverse sides, in other words, doing this is completely fine, as the author will be propagated to the new books:

```ts
em.create(Author, {
  name: '...',
  books: [{ title: 'foo' }], // Book.author is required, but will be propagated from the upper scope
});
```

I managed to find a way to mark those properties as optional automatically, based on the parent scope - in other words, since we are processing contents of `Author.books`, we will make all properties of type `Author` in the `Author.books` as optional. It's not perfect, as it will mark every `Author` property as optional, not just the one that is an inverse of `Author.books`, but I feel like it's the tradeoff we need to make.
